File: backend/app/tools.py
"""Agent tools: the bridge between the LLM and the deterministic engine.

Each tool is a small, well-typed function that pulls data (data.py) and runs
scoring (scoring.py). The LLM may only obtain numbers by calling these tools —
it never invents figures. Every tool returns a `ToolOutput` carrying:

  * result       — compact JSON the LLM reads to write its answer
  * structured   — optional table/chart payload for the UI (per the API contract)
  * assumptions  — the deterministic assumptions behind the numbers
  * uncertainty  — honest caveats to surface to the user

`TOOL_SCHEMAS` is the Anthropic tool-definition list; `run_tool` dispatches a
call by name.
"""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any

from . import data, scoring


@dataclass
class ToolOutput:
    result: dict[str, Any]
    structured: dict[str, Any] | None = None
    assumptions: list[str] = field(default_factory=list)
    uncertainty: str | None = None


# --------------------------------------------------------------------------- #
# Tool implementations
# --------------------------------------------------------------------------- #

def _score_columns(metric: str) -> str:
    return {
        "congestion": "Congestion score",
        "unmet_demand": "Unmet-demand score",
        "expansion": "Expansion score",
    }[metric]


def _unknown(iata: str) -> ToolOutput:
    return ToolOutput(
        result={
            "error": f"'{iata}' is not in the dataset.",
            "known_airports": data.known_iatas(),
        }
    )


def list_airports(scope: str | None = None, states: list[str] | None = None) -> ToolOutput:
    airports = data.airports_in_scope(scope, states)
    rows = [
        {"iata": a["iata"], "name": a["name"], "city": a["city"], "state": a["state"]}
        for a in airports
    ]
    return ToolOutput(
        result={"count": len(rows), "airports": rows, "scope": scope or "all"},
        assumptions=[f"Dataset scope is limited to {len(data.all_airports())} U.S. airports."],
    )


def airport_profile(iata: str) -> ToolOutput:
    airport = data.get_airport(iata)
    if not airport:
        return _unknown(iata)

    congestion = scoring.congestion_score(airport)
    unmet = scoring.unmet_demand_score(airport)
    expansion = scoring.expansion_score(airport)

    return ToolOutput(
        result={
            "iata": airport["iata"],
            "name": airport["name"],
            "city": airport["city"],
            "state": airport["state"],
            "metrics": {
                "annual_passengers": airport["annual_passengers"],
                "pax_growth_yoy": airport["pax_growth_yoy"],
                "load_factor": airport["load_factor"],
                "avg_dep_delay_min": airport["avg_dep_delay_min"],
                "delayed_share": airport["delayed_share"],
                "cancel_rate": airport["cancel_rate"],
                "annual_departures": airport["annual_departures"],
                "runways": airport["runways"],
            },
            "haul_mix_pct": scoring.haul_breakdown(airport),
            "scores": {
                "congestion": congestion.as_dict(),
                "unmet_demand": unmet.as_dict(),
                "expansion": expansion.as_dict(),
            },
        },
        assumptions=scoring.EXPANSION_ASSUMPTIONS[:1],
    )


def rank_airports(scope: str | None = "all", metric: str = "expansion", top_n: int = 5) -> ToolOutput:
    if metric not in scoring.METRIC_SCORERS:
        return ToolOutput(result={"error": f"Unknown metric '{metric}'.",
                                  "valid_metrics": list(scoring.METRIC_SCORERS)})

    airports = data.airports_in_scope(scope)
    if not airports:
        return ToolOutput(result={"error": f"No airports found for scope '{scope}'."})

    scorer = scoring.METRIC_SCORERS[metric]
    ranked = sorted(
        ((a, scorer(a)) for a in airports),
        key=lambda pair: pair[1].score,
        reverse=True,
    )[:max(1, top_n)]

    columns = ["Rank", "Airport", _score_columns(metric), "Growth YoY", "Load factor", "Avg delay (min)"]
    rows = [
        [
            i + 1,
            f'{a["iata"]} — {a["city"]}',
            round(res.score, 1),
            f'{a["pax_growth_yoy"] * 100:.1f}%',
            round(a["load_factor"], 2),
            a["avg_dep_delay_min"],
        ]
        for i, (a, res) in enumerate(ranked)
    ]

    sample = scorer(ranked[0][0])
    return ToolOutput(
        result={
            "metric": metric,
            "scope": scope,
            "ranking": [
                {"iata": a["iata"], "city": a["city"], "score": round(res.score, 1),
                 "components": res.as_dict()["components"]}
                for a, res in ranked
            ],
        },
        structured={"kind": "ranking", "columns": columns, "rows": rows},
        assumptions=sample.assumptions,
        uncertainty=sample.uncertainty,
    )


def compare_airports(iatas: list[str], metric: str = "congestion") -> ToolOutput:
    if metric not in scoring.METRIC_SCORERS:
        return ToolOutput(result={"error": f"Unknown metric '{metric}'.",
                                  "valid_metrics": list(scoring.METRIC_SCORERS)})

    resolved = [(code, data.get_airport(code)) for code in iatas]
    missing = [code for code, a in resolved if a is None]
    if missing:
        return ToolOutput(result={"error": f"Unknown airport(s): {', '.join(missing)}.",
                                  "known_airports": data.known_iatas()})

    scorer = scoring.METRIC_SCORERS[metric]
    scored = [(a, scorer(a)) for _, a in resolved]

    columns = ["Airport", _score_columns(metric), "Avg delay (min)", "Delayed >15m", "Load factor", "Cancel rate"]
    rows = [
        [
            f'{a["iata"]} — {a["city"]}',
            round(res.score, 1),
            a["avg_dep_delay_min"],
            f'{a["delayed_share"] * 100:.0f}%',
            round(a["load_factor"], 2),
            f'{a["cancel_rate"] * 100:.1f}%',
        ]
        for a, res in scored
    ]

    return ToolOutput(
        result={
            "metric": metric,
            "comparison": [
                {"iata": a["iata"], "score": round(res.score, 1),
                 "components": res.as_dict()["components"]}
                for a, res in scored
            ],
        },
        structured={"kind": "comparison", "columns": columns, "rows": rows},
        assumptions=scored[0][1].assumptions,
        uncertainty=scored[0][1].uncertainty,
    )


def long_haul_breakdown(iata: str) -> ToolOutput:
    airport = data.get_airport(iata)
    if not airport:
        return _unknown(iata)

    mix = scoring.haul_breakdown(airport)
    thresholds = data.haul_thresholds()
    columns = ["Haul type", "Share of departures"]
    rows = [
        [f'Short (<{thresholds["short_max"]} mi)', f'{mix["short"]}%'],
        [f'Medium ({thresholds["short_max"]}-{thresholds["long_min"]} mi)', f'{mix["medium"]}%'],
        [f'Long (>{thresholds["long_min"]} mi)', f'{mix["long"]}%'],
    ]
    return ToolOutput(
        result={"iata": airport["iata"], "name": airport["name"], "haul_mix_pct": mix,
                "thresholds_miles": thresholds},
        structured={"kind": "breakdown", "columns": columns, "rows": rows},
        assumptions=[
            "Haul class is by great-circle route distance from T-100 segment data: "
            f'short <{thresholds["short_max"]} mi, medium {thresholds["short_max"]}-'
            f'{thresholds["long_min"]} mi, long >{thresholds["long_min"]} mi.',
            "Percentages are shares of departures performed, not passengers or seats.",
        ],
    )


def unmet_demand(iata: str) -> ToolOutput:
    airport = data.get_airport(iata)
    if not airport:
        return _unknown(iata)

    res = scoring.unmet_demand_score(airport)
    comp = res.as_dict()["components"]
    columns = ["Driver", "Contribution (0-1)"]
    rows = [
        ["Load-factor pressure", comp["load_pressure"]],
        ["Congestion (delay/throughput)", comp["congestion"]],
        ["Growth vs. capacity", comp["growth_vs_capacity"]],
    ]
    return ToolOutput(
        result={
            "iata": airport["iata"],
            "name": airport["name"],
            "unmet_demand_score": round(res.score, 1),
            "drivers": comp,
            "context": {
                "load_factor": airport["load_factor"],
                "pax_growth_yoy": airport["pax_growth_yoy"],
                "runways": airport["runways"],
                "avg_dep_delay_min": airport["avg_dep_delay_min"],
            },
        },
        structured={"kind": "metric", "columns": columns, "rows": rows},
        assumptions=res.assumptions,
        uncertainty=res.uncertainty,
    )


# --------------------------------------------------------------------------- #
# Dispatch + schemas
# --------------------------------------------------------------------------- #

_DISPATCH = {
    "list_airports": list_airports,
    "airport_profile": airport_profile,
    "rank_airports": rank_airports,
    "compare_airports": compare_airports,
    "long_haul_breakdown": long_haul_breakdown,
    "unmet_demand": unmet_demand,
}


def run_tool(name: str, tool_input: dict[str, Any]) -> ToolOutput:
    """Execute a tool by name with the LLM-supplied arguments."""
    func = _DISPATCH.get(name)
    if not func:
        return ToolOutput(result={"error": f"Unknown tool '{name}'."})
    try:
        return func(**tool_input)
    except TypeError as exc:
        return ToolOutput(result={"error": f"Bad arguments for '{name}': {exc}"})


TOOL_SCHEMAS: list[dict[str, Any]] = [
    {
        "name": "list_airports",
        "description": "List airports available in the dataset, optionally filtered by "
                       "scope ('new_england', 'all', or a 2-letter state code) or an explicit "
                       "list of state codes. Use this to discover which airports exist before ranking.",
        "input_schema": {
            "type": "object",
            "properties": {
                "scope": {"type": "string", "description": "'new_england', 'all', or a state code like 'CA'."},
                "states": {"type": "array", "items": {"type": "string"}, "description": "Explicit state codes; overrides scope."},
            },
        },
    },
    {
        "name": "airport_profile",
        "description": "Full metric profile for one airport by IATA code (e.g. 'SFO'): traffic, "
                       "delays, load factor, haul mix, and all deterministic scores.",
        "input_schema": {
            "type": "object",
            "properties": {"iata": {"type": "string", "description": "IATA code, e.g. 'BOS'."}},
            "required": ["iata"],
        },
    },
    {
        "name": "rank_airports",
        "description": "Rank airports within a scope by a deterministic metric. Use for questions "
                       "like 'best candidates for terminal expansion'. metric='expansion' for "
                       "expansion candidates, 'congestion', or 'unmet_demand'.",
        "input_schema": {
            "type": "object",
            "properties": {
                "scope": {"type": "string", "description": "'new_england', 'all', or a state code."},
                "metric": {"type": "string", "enum": ["expansion", "congestion", "unmet_demand"]},
                "top_n": {"type": "integer", "description": "How many to return (default 5)."},
            },
            "required": ["metric"],
        },
    },
    {
        "name": "compare_airports",
        "description": "Compare two or more airports (by IATA code) on a metric. Use for "
                       "'compare X and Y congestion'. metric='congestion', 'unmet_demand', or 'expansion'.",
        "input_schema": {
            "type": "object",
            "properties": {
                "iatas": {"type": "array", "items": {"type": "string"}, "description": "IATA codes, e.g. ['LAX','SNA']."},
                "metric": {"type": "string", "enum": ["congestion", "unmet_demand", "expansion"]},
            },
            "required": ["iatas"],
        },
    },
    {
        "name": "long_haul_breakdown",
        "description": "Short/medium/long-haul share of departures for one airport by IATA code. "
                       "Use for 'percentage of long-haul flights out of X'.",
        "input_schema": {
            "type": "object",
            "properties": {"iata": {"type": "string"}},
            "required": ["iata"],
        },
    },
    {
        "name": "unmet_demand",
        "description": "Estimate unmet flight demand for one airport by IATA code, with the "
                       "drivers behind it. Use for 'unmet demand at X and why'.",
        "input_schema": {
            "type": "object",
            "properties": {"iata": {"type": "string"}},
            "required": ["iata"],
        },
    },
]

File: lib/tools.ts
/**
 * Agent tools: the bridge between the LLM and the deterministic engine.
 *
 * Each tool pulls data (data.ts) and runs scoring (scoring.ts). The LLM may only
 * obtain numbers by calling these tools — it never invents figures. Every tool
 * returns a `ToolOutput` carrying:
 *   - result       compact JSON the LLM reads to write its answer
 *   - structured   optional table/chart payload for the UI (per the API contract)
 *   - assumptions  the deterministic assumptions behind the numbers
 *   - uncertainty  honest caveats to surface to the user
 *
 * `TOOL_SCHEMAS` is the provider-neutral tool-definition list; `runTool` dispatches by name.
 */

import * as data from "./data";
import type { Airport } from "./data";
import {
  METRIC_SCORERS,
  haulBreakdown,
  congestionScore,
  unmetDemandScore,
  expansionScore,
  roundResult,
  type MetricKey,
} from "./scoring";
import {
  ASSUMPTIONS,
  haulAssumptions,
  liveFlightsAssumptions,
  LIVE_FLIGHTS_UNCERTAINTY,
} from "./i18n";
import { flightsNearAirport } from "./liveflights";
import type { StructuredResult } from "./types";

export interface ToolOutput {
  result: Record<string, unknown>;
  structured?: StructuredResult | null;
  assumptions?: string[];
  uncertainty?: string | null;
}

const SCORE_COLUMN: Record<MetricKey, string> = {
  congestion: "Congestion score",
  unmet_demand: "Unmet-demand score",
  expansion: "Expansion score",
};

const round1 = (n: number): number => Math.round(n * 10) / 10;
const pct = (frac: number, dp = 0): string => `${(frac * 100).toFixed(dp)}%`;

function unknown(iata: string): ToolOutput {
  return { result: { error: `'${iata}' is not in the dataset.`, known_airports: data.knownIatas() } };
}

// --------------------------------------------------------------------------- //
// Tool implementations
// --------------------------------------------------------------------------- //

export function airportProfile(iata: string): ToolOutput {
  const a = data.getAirport(iata);
  if (!a) return unknown(iata);
  return {
    result: {
      iata: a.iata,
      name: a.name,
      city: a.city,
      state: a.state,
      metrics: {
        annual_passengers: a.annual_passengers,
        pax_growth_yoy: a.pax_growth_yoy,
        load_factor: a.load_factor,
        avg_dep_delay_min: a.avg_dep_delay_min,
        delayed_share: a.delayed_share,
        cancel_rate: a.cancel_rate,
        annual_departures: a.annual_departures,
        runways: a.runways,
      },
      haul_mix_pct: haulBreakdown(a),
      scores: {
        congestion: roundResult(congestionScore(a)),
        unmet_demand: roundResult(unmetDemandScore(a)),
        expansion: roundResult(expansionScore(a)),
      },
    },
    assumptions: [ASSUMPTIONS.expansion[0]],
  };
}

export function rankAirports(
  scope: string = "all",
  metric: MetricKey = "expansion",
  topN = 5,
): ToolOutput {
  if (!(metric in METRIC_SCORERS)) {
    return { result: { error: `Unknown metric '${metric}'.`, valid_metrics: Object.keys(METRIC_SCORERS) } };
  }
  const airports = data.airportsInScope(scope);
  if (airports.length === 0) return { result: { error: `No airports found for scope '${scope}'.` } };

  const scorer = METRIC_SCORERS[metric];
  const ranked = airports
    .map((a) => ({ a, res: scorer(a) }))
    .sort((x, y) => y.res.score - x.res.score)
    .slice(0, Math.max(1, topN));

  const columns = ["Rank", "Airport", SCORE_COLUMN[metric], "Growth YoY", "Load factor", "Avg delay (min)"];
  const rows = ranked.map(({ a, res }, i) => [
    i + 1,
    `${a.iata} — ${a.city}`,
    round1(res.score),
    pct(a.pax_growth_yoy, 1),
    Number(a.load_factor.toFixed(2)),
    a.avg_dep_delay_min,
  ]);

  const sample = ranked[0].res;
  return {
    result: {
      metric,
      scope,
      note: "score is a 0-100 composite; the other fields are the real underlying metrics.",
      ranking: ranked.map(({ a, res }) => ({
        iata: a.iata,
        city: a.city,
        score: round1(res.score),
        pax_growth_yoy_pct: Number((a.pax_growth_yoy * 100).toFixed(1)),
        load_factor: a.load_factor,
        avg_dep_delay_min: a.avg_dep_delay_min,
        annual_passengers: a.annual_passengers,
      })),
    },
    structured: { kind: "ranking", columns, rows },
    assumptions: sample.assumptions,
    uncertainty: sample.uncertainty,
  };
}

export function compareAirports(
  iatas: string[],
  metric: MetricKey = "congestion",
): ToolOutput {
  if (!(metric in METRIC_SCORERS)) {
    return { result: { error: `Unknown metric '${metric}'.`, valid_metrics: Object.keys(METRIC_SCORERS) } };
  }
  const resolved = iatas.map((code) => ({ code, a: data.getAirport(code) }));
  const missing = resolved.filter((r) => !r.a).map((r) => r.code);
  if (missing.length > 0) {
    return { result: { error: `Unknown airport(s): ${missing.join(", ")}.`, known_airports: data.knownIatas() } };
  }

  const scorer = METRIC_SCORERS[metric];
  const scored = resolved.map((r) => ({ a: r.a as Airport, res: scorer(r.a as Airport) }));

  const columns = [
    "Airport",
    SCORE_COLUMN[metric],
    "Avg delay (min)",
    "Delayed >15m",
    "Load factor",
    "Cancel rate",
  ];
  const rows = scored.map(({ a, res }) => [
    `${a.iata} — ${a.city}`,
    round1(res.score),
    a.avg_dep_delay_min,
    pct(a.delayed_share),
    Number(a.load_factor.toFixed(2)),
    pct(a.cancel_rate, 1),
  ]);

  return {
    result: {
      metric,
      note: "score is a 0-100 composite; the other fields are the real underlying metrics.",
      comparison: scored.map(({ a, res }) => ({
        iata: a.iata,
        city: a.city,
        score: round1(res.score),
        avg_dep_delay_min: a.avg_dep_delay_min,
        pct_flights_delayed_over_15min: Math.round(a.delayed_share * 100),
        cancel_rate_pct: Number((a.cancel_rate * 100).toFixed(1)),
        load_factor: a.load_factor,
      })),
    },
    structured: { kind: "comparison", columns, rows },
    assumptions: scored[0].res.assumptions,
    uncertainty: scored[0].res.uncertainty,
  };
}

export function longHaulBreakdown(iata: string): ToolOutput {
  const a = data.getAirport(iata);
  if (!a) return unknown(iata);
  const mix = haulBreakdown(a);
  const t = data.haulThresholds();
  const columns = ["Haul type", "Share of departures"];
  const rows = [
    [`Short (<${t.short_max} mi)`, `${mix.short}%`],
    [`Medium (${t.short_max}-${t.long_min} mi)`, `${mix.medium}%`],
    [`Long (>${t.long_min} mi)`, `${mix.long}%`],
  ];
  return {
    result: { iata: a.iata, name: a.name, haul_mix_pct: mix, thresholds_miles: t },
    structured: { kind: "breakdown", columns, rows },
    assumptions: haulAssumptions(t),
  };
}

export function unmetDemand(iata: string): ToolOutput {
  const a = data.getAirport(iata);
  if (!a) return unknown(iata);
  const res = unmetDemandScore(a);
  const comp = roundResult(res).components;
  const columns = ["Driver", "Contribution (0-1)"];
  const rows = [
    ["Load-factor pressure", comp.load_pressure],
    ["Congestion (delay/throughput)", comp.congestion],
    ["Growth vs. capacity", comp.growth_vs_capacity],
  ];
  return {
    result: {
      iata: a.iata,
      name: a.name,
      unmet_demand_score: round1(res.score),
      drivers: comp,
      context: {
        load_factor: a.load_factor,
        pax_growth_yoy: a.pax_growth_yoy,
        runways: a.runways,
        avg_dep_delay_min: a.avg_dep_delay_min,
      },
    },
    structured: { kind: "metric", columns, rows },
    assumptions: res.assumptions,
    uncertainty: res.uncertainty,
  };
}

export async function liveFlights(iata: string): Promise<ToolOutput> {
  const a = data.getAirport(iata);
  if (!a) return unknown(iata);

  try {
    const live = await flightsNearAirport(a.lat, a.lon);
    const columns = ["Live activity (now)", "Count"];
    const rows = [
      ["Airborne nearby", live.airborne],
      ["On the ground", live.onGround],
      ["Total in area", live.total],
    ];
    return {
      result: {
        iata: a.iata,
        name: a.name,
        available: true,
        airborne_nearby: live.airborne,
        on_ground: live.onGround,
        total_in_area: live.total,
        sample_callsigns: live.sample.map((f) => f.callsign),
      },
      structured: { kind: "metric", columns, rows },
      assumptions: liveFlightsAssumptions(),
      uncertainty: LIVE_FLIGHTS_UNCERTAINTY,
    };
  } catch (err) {
    return {
      result: {
        iata: a.iata,
        available: false,
        message: `Couldn't reach the live flight feed: ${(err as Error).message}`,
      },
    };
  }
}

// --------------------------------------------------------------------------- //
// Dispatch + schemas
// --------------------------------------------------------------------------- //

type ToolFn = (args: Record<string, unknown>) => ToolOutput | Promise<ToolOutput>;

const DISPATCH: Record<string, ToolFn> = {
  airport_profile: (a) => airportProfile(a.iata as string),
  rank_airports: (a) =>
    rankAirports(a.scope as string | undefined, (a.metric as MetricKey) ?? "expansion", (a.top_n as number) ?? 5),
  compare_airports: (a) => compareAirports((a.iatas as string[]) ?? [], (a.metric as MetricKey) ?? "congestion"),
  long_haul_breakdown: (a) => longHaulBreakdown(a.iata as string),
  unmet_demand: (a) => unmetDemand(a.iata as string),
  live_flights: (a) => liveFlights(a.iata as string),
};

/** Execute a tool by name with the LLM-supplied arguments. */
export async function runTool(name: string, input: Record<string, unknown>): Promise<ToolOutput> {
  const fn = DISPATCH[name];
  if (!fn) return { result: { error: `Unknown tool '${name}'.` } };
  try {
    return await fn(input ?? {});
  } catch (err) {
    return { result: { error: `Bad arguments for '${name}': ${(err as Error).message}` } };
  }
}

/** Provider-neutral tool definition (mapped to the LLM SDK's format in agent.ts). */
export interface ToolSchema {
  name: string;
  description: string;
  input_schema: Record<string, unknown>;
}

export const TOOL_SCHEMAS: ToolSchema[] = [
  {
    name: "airport_profile",
    description:
      "Full metric profile for one airport by IATA code (e.g. 'SFO'): traffic, delays, " +
      "load factor, haul mix, and all deterministic scores.",
    input_schema: {
      type: "object",
      properties: { iata: { type: "string", description: "IATA code, e.g. 'BOS'." } },
      required: ["iata"],
    },
  },
  {
    name: "rank_airports",
    description:
      "Rank airports within a scope by a deterministic metric. Use for questions like " +
      "'best candidates for terminal expansion'. metric='expansion' for expansion " +
      "candidates, 'congestion', or 'unmet_demand'.",
    input_schema: {
      type: "object",
      properties: {
        scope: { type: "string", description: "'new_england', 'all', or a state code." },
        metric: { type: "string", enum: ["expansion", "congestion", "unmet_demand"] },
        top_n: { type: "integer", description: "How many to return (default 5)." },
      },
      required: ["metric"],
    },
  },
  {
    name: "compare_airports",
    description:
      "Compare two or more airports (by IATA code) on a metric. Use for 'compare X and Y " +
      "congestion'. metric='congestion', 'unmet_demand', or 'expansion'.",
    input_schema: {
      type: "object",
      properties: {
        iatas: { type: "array", items: { type: "string" }, description: "IATA codes, e.g. ['LAX','SNA']." },
        metric: { type: "string", enum: ["congestion", "unmet_demand", "expansion"] },
      },
      required: ["iatas"],
    },
  },
  {
    name: "long_haul_breakdown",
    description:
      "Short/medium/long-haul share of departures for one airport by IATA code. Use for " +
      "'percentage of long-haul flights out of X'.",
    input_schema: {
      type: "object",
      properties: { iata: { type: "string" } },
      required: ["iata"],
    },
  },
  {
    name: "unmet_demand",
    description:
      "Estimate unmet flight demand for one airport by IATA code, with the drivers behind " +
      "it. Use for 'unmet demand at X and why'.",
    input_schema: {
      type: "object",
      properties: { iata: { type: "string" } },
      required: ["iata"],
    },
  },
  {
    name: "live_flights",
    description:
      "Real-time count of aircraft in the air right now near an airport (by IATA code), " +
      "from live ADS-B data. Use ONLY for questions about current / live / right-now " +
      "flight activity — not for historical traffic, rankings, or scores.",
    input_schema: {
      type: "object",
      properties: { iata: { type: "string" } },
      required: ["iata"],
    },
  },
];
